Compatibility rework

rankynbass opened the pull request:
This replaces the Dxvk Settings rework - #17.

Here's the compatibilitytools rework I mentioned a few weeks ago. This moves most of the bits that will change over to the XL.Core side, such as specific versions of wine or dxvk, where they will be stored, and what environment variables need to be set for each one.

I also moved the MacVideoFix over to XL.Core. I figure individual implementations of fixes can be done on the XL.Core side, and then passed to GameFixApply object.

Depends on https://github.com/goatcorp/FFXIVQuickLauncher/pull/1348

And here's a link to a flatpak build if you want to test:
https://github.com/rankynbass/XIVLauncher.Core/releases/edit/compatrework-v1

Install with `flatpak install xivlauncher-rework.flatpak`, run with `flatpak run dev.goats.testing.xivlauncher`.

Blooym commented on src/XIVLauncher.Core/WineManager.cs:
```diff
@@ -0,0 +1,97 @@
+    Wine7_10,
+}
+
+public static class WineManager
+{
+#if WINE_XIV_ARCH_LINUX
+    private const string DISTRO = "arch";
+#elif WINE_XIV_FEDORA_LINUX
+    private const string DISTRO = "fedora";
+#else
+    private const string DISTRO = "ubuntu";
+#endif
```
Would it be possible for us to determine this without setting flags at build time so we only need to produce a single artifact?

rankynbass replied:
Actually, yes. On just about any system, we can read the /etc/os-release file, and the first line will be NAME="some distro". So we could search for Arch and Fedora/SUSE, and then give ubuntu if it doesn't match those.

Blooym replied:
Perfect, would be nice if we could get that change in with this if possible

rankynbass replied:
Alright, I've added a function for this. Basically, fedora has an /etc/fedora-release file, arch derivatives have an /etc/arch-release file, and everyone else can get the ubuntu release.

Blooym commented on src/XIVLauncher.Core/GameFixes/MacVideoFix.cs:
```diff
@@ -0,0 +1,59 @@
+
+public class MacVideoFix : GameFix
+{
+    private const string MAC_ZIP_URL = "https://mac-dl.ffxiv.com/cw/finalfantasyxiv-1.1.2.zip";
```
Should we have this point to somewhere in our repo (such as `assets/mac_fix_url`) so that we can update it without pushing an entire launcher update. It would make it easier for us and 3rd party packagers as if this breaks often we can fix it faster.

Blooym reviewed (approved):
Happy with the changes on the XLCore side to go forward once the submodule changes are in.

Blooym commented on src/XIVLauncher.Core/Program.cs:
```diff
@@ -44,7 +44,9 @@ class Program
     private static readonly Vector3 clearColor = new(0.1f, 0.1f, 0.1f);
     private static bool showImGuiDemoWindow = true;
-
+    public static string Distro { get; private set; }
```
Might be an idea to make this an enum so we're more aware of the valid values for it, could then just assign an attribute of its url value so we cna download the wine version from it using that attribute name.

Blooym commented on src/XIVLauncher.Core/Program.cs:
```diff
@@ -383,6 +385,61 @@ private static ISecretProvider GetSecretProvider(Storage storage)
+                    if (keyValue[1].Contains("arch"))
+                        distro = "arch";
+                    if (keyValue[1].Contains("fedora"))
+                        distro = "fedora";
```
I don't appear to have `ID_LIKE` on Fedora Silverblue, not sure if it'd get properly set to Fedora because of this?

rankynbass replied:
Are you actually using the rpm on silverblue? I thought the idea with that was to use flatpaks. In any case, if you could post the contents of your /etc/os-release file, I'll see about adding it.

Blooym replied:
I use both depending on which one is causing me less problems at the time.

```
NAME="Fedora Linux"
VERSION="38.20230630.0 (Silverblue)"
ID=fedora
VERSION_ID=38
VERSION_CODENAME=""
PLATFORM_ID="platform:f38"
PRETTY_NAME="Fedora Linux 38.20230630.0 (Silverblue)"
DEFAULT_HOSTNAME="fedora"
REDHAT_BUGZILLA_PRODUCT="Fedora"
REDHAT_BUGZILLA_PRODUCT_VERSION=38
REDHAT_SUPPORT_PRODUCT="Fedora"
REDHAT_SUPPORT_PRODUCT_VERSION=38
VARIANT="Silverblue"
VARIANT_ID=silverblue
. . . removed un-needed values
```

Blooym reviewed (commented):
Overall happy with the changes to detect distribution, just have some small questions/thoughts about it

marzent commented:
Honestly not sure about moving game fixes to xlcore wholesale, maybe adding a parameter to it for the currently hosted version by SE would be cleaner, ultimately something like `$"https://mac-dl.ffxiv.com/cw/finalfantasyxiv-{currentSeMacWrapperVersion}.zip"`.

Then it still possible to add in dynamic version information fetching on the xlcore side later.

Blooym commented:
> Honestly not sure about moving game fixes to xlcore wholesale, maybe adding a parameter to it for the currently hosted version by SE would be cleaner, ultimately something like `$"https://mac-dl.ffxiv.com/cw/finalfantasyxiv-{currentSeMacWrapperVersion}.zip"`.
> 
> Then it still possible to add in dynamic version information fetching on the xlcore side later.

The way gamefixes is setup here doesn't seem like too much of a problem to me since it allows passing it in from either side & it makes it easier to adjust fixes that aren't going to apply to the main repo anyway

marzent commented:
>  it makes it easier to adjust fixes

I agree with that and that was probably the main motivation behind this change... 

However xlcore is mostly a frontend for XIVLauncher.Common and this change muddies the waters a bit (there is for example a special exception in the game repair logic for these video files, you could add validation there in the future as well, but this is gonna become difficult to maintain by splitting parts of the codebase for convenience and not architecture)

Ultimately this is a similar argument to https://github.com/goatcorp/XIVLauncher.Core/pull/10

Blooym commented:
> However xlcore is mostly a frontend for XIVLauncher.Common and this change muddies the waters a bit (there is for example a special exception in the game repair logic for these video files, you could add validation there in the future as well, but this is gonna become difficult to maintain by splitting parts of the codebase for convenience and not architecture)

I'm totally happy to keep things in XIVLauncher.Common but it would be nice to have an easier path to getting changes made in XIVLauncher.Common.Unix as right now any functional change we want made has to go through it.

marzent commented:
Yeah XIVLauncher.Common has become a bit of a bottleneck for functional changes...

Having said that (slightly off-topic again) I think the correct video URL can be extracted from SEs sparkle feed, so this won't really need any adjustments in the future.

rankynbass commented:
So should I revert this? We could also have GameApplyFixes to either accept an array/list as a contstructor arg (as I already did), or add a function to add fixes one at a time. Then you could just do something like
```
var gameFixApply = new GameFixApply();
gameFixApply.Add(new MacVideoFix(...));
```

marzent commented:
Personally I am not a huge fan of adding game fixes from the xlcore side (as in xlcore holds their implementation), but that is a matter of taste.
Haven't had the time to look over everything and test in detail but it looks good on first glance (and thanks for keeping the wine and dxvk runners in XIVLauncher.Common).
Can you maybe rebase both this and https://github.com/goatcorp/FFXIVQuickLauncher/pull/1348 (and drop 8fd716db41069f59575d25a0199c0c0db0a4267b)

rankynbass commented:
Okay, reverted the video fix move.

rankynbass commented on src/XIVLauncher.Core/Components/MainPage/MainPage.cs:
```diff
@@ -761,7 +761,17 @@ public async Task<Process> StartGameAndAddon(Launcher.LoginResult loginResult, b
+                if (Program.Config.DxvkVersion == DxvkVersion.Disabled)
+                {
+                    if (Program.Config.WineD3DUseVK ?? false)
+                        Program.CompatibilityTools.AddRegistryKey("HKEY_CURRENT_USER\\Software\\Wine\\Direct3D", "renderer", "vulkan");
```
This one isn't really that important, and can easily be removed. It's just an option to use WineD3D in Vulkan mode, which is not terribly stable to begin with. As for putting it somewhere else, the only other place is in UnixGameRunner, and it'd still probably lock up for a few seconds. I'm just going to get rid of the WineD3D vulkan stuff. Chances are the only reason someone would want to use WineD3D is to use openGL anyway.

rankynbass commented on src/XIVLauncher.Core/Distro.cs:
```diff
@@ -0,0 +1,76 @@
+
+namespace XIVLauncher.Core;
+
+public enum WinePackage
```
It's actually used in the WineManager to determine which wine package to download. I've renamed to DistroPackage now.

rankynbass commented on src/XIVLauncher.Core/WineManager.cs:
```diff
@@ -0,0 +1,96 @@
+        return new WineRunner(winepath, wineargs, folder, url, Program.storage.Root.FullName, env);
+    }
+
+    private static string GetDistro()
```
We are, actually. This is leftover code that I have now removed.

rankynbass commented on src/XIVLauncher.Core/DxvkManager.cs:
```diff
@@ -0,0 +1,208 @@
+                env.Add("DXVK_HUD","0");
+                env.Add("MANGOHUD","1");
+
+                if (mangoHudConfig is null)
```
Moved the hud stuff to its own class. There's also now a function that checks for mangohud being installed, and will set hudtype to none if somehow the user ends up manually setting mangohud in the .ini file without having it installed. DxvkSettingsTab already will not let you save settings in that case.

rankynbass commented on src/XIVLauncher.Core/Components/SettingsPage/Tabs/SettingsTabDxvk.cs:
```diff
@@ -0,0 +1,100 @@
+                    return null;
+                },
+            },
+            new SettingsEntry<string>("MangoHud Custom Path", "Set a custom path for MangoHud config file.", () => Program.Config.DxvkMangoCustom, s => Program.Config.DxvkMangoCustom = s)
```
For the moment, there's a function that looks for specific paths where Linux installs MangoHud, and it won't let you save your selection if it doesn't find it (it will show an error). I think the whole DXVK tab is really linux only. Only the Disabled/WineD3D setting would work on MacOS atm (if that), and only if you manually provided a wine release. Maybe there's DXVK for freebsd? Not sure

marzent replied:
DXVK is a PE only binary that translates Windows dx calls into Windows Vulkan calls (simplified), so it will run on any platform that supports the necessary Windows Vulkan feature level which is Wine on Linux, FreeBSD, macOS (if the native unix implementation has all needed extensions) and even on Windows itself (where it is sometimes used to provide better frame pacing, or just in general better performance that the native Windows dx driver)

rankynbass commented:
So, in no particular order:
- WineD3D vulkan was removed. Adding a reg key can be slow, and this option is quite buggy with wine-xiv.
- Cleaned up the Distro stuff to make it more platform independent.
- Cleaned up some variable names and function names to be less confusing (hopefully). Deleted some old leftover functions.
- Moved all the WINEDLLOVERRIDES stuff to XIVLauncher.Common.Unix
- Moved the WineManger, DxvkManager to XIVLauncher.Core.UnixCompatibility, since technically this can be compiled for Windows. Also split out the Hud stuff to a separate class just to make it a little easier to manage the code.

rankynbass commented on src/XIVLauncher.Core/Program.cs:
```diff
@@ -319,11 +324,20 @@ public static void CreateCompatToolsInstance()
+        var wine = WineManager.Initialize();
+        var dxvk = DxvkManager.Initialize();
+        var wineoverride = "msquic=,mscoree=n,b;";
+        var wineenv = new Dictionary<string, string>();
```
Fixed. Now it's all in CompatibilityTools

rankynbass commented on src/XIVLauncher.Core/Components/SettingsPage/Tabs/SettingsTabDebug.cs:
```diff
@@ -12,7 +12,7 @@ public override void Draw()
         ImGui.TextUnformatted("Generic Information");
         ImGui.Separator();
-        ImGui.TextUnformatted($"Operating System: {Environment.OSVersion}");
+        ImGui.TextUnformatted($"Operating System: {Distro.Name} - {Environment.OSVersion}");
```
I might go through and fix all of this at some point; do the full initialization of OS info in the Distro object (maybe rename it), and then change all the Environment.OSVersion.Platform checks to refer to the XIVLauncher.Common Platform value set in said Distro object. But I think there's enough going on atm. For now, I've added in some checks to make sure the Linux-specific stuff only shows up if it's actually on Linux.

rankynbass commented:
Closing and replacing with https://github.com/goatcorp/XIVLauncher.Core/pull/64
